ls1intum/Artemis
# `Assessment`: Allow instructors to export course scores to an Excel file

### Checklist
#### General


#### Client


### Motivation and Context


Currently Artemis only allows to export course scores to a CSV file. When a user specifies another decimal separator than their system's setting and then imports the CSV file into Excel, values are interpreted differently (e.g. Custom Decimal Separator: '.', User system set to german: ',' -> Generated csv contains numbers in the format 1.4 -> Excel interprets this as a date (1. April)).
In addition, allowing to export directly to an Excel file improves the usability for technical inexperienced users.


### Description

The introduced changes now allow an instructor to download course scores directly as an Excel file by utilizing the library [SheetJS](https://github.com/SheetJS/sheetjs).

### Steps for Testing

Prerequisites:

- 1 Course
- 1 Student
- 1 Exercise, on which the student participated

1. Log in to Artemis
2. Navigate to Course Administration, select the course, then Scores
3. Click the `Export` button
4. A modal opens, where you can select the file type used for the export (Excel<default> or CSV).
5. Select Excel (which should be selected initially) and click `Export as Excel`
6. Open the downloaded Excel file and make sure the rows and their values are equivalent to the ones displayed in Artemis
7. In addition, also verify that the export to a CSV file still works as expected.

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
| course-scores-csv-row-builder.ts | 100% | 100% |
| course-scores-excel-row-builder.ts | 100% | 100% |
| course-scores-row-builder.ts | 96% | 100% |
| course-scores.component.ts | 73.52% | 93.17% |
| export-modal.component.ts | 100% | 100% |


### Screenshots


![Excel_Export](https://user-images.githubusercontent.com/9954674/165321171-35a86246-0e79-4de6-935a-6515f30bf567.gif)
